Put in the changes to jump ID3 tags.

leftCoast opened the pull request:
Apple, and I think windows, has been putting massive ID3 meta tags in
the front of .mp3 files. They typically take more than 20 sec. to start
playing as the player rolls through this stuff. This fix jumps the file
pointer beyond the ID3 tag for a speedy quick startup.

TheNitek commented:
Any chance this gets merged? It's a really huge improvement in many cases!

ladyada commented:
we want to fix stability changes for now, please use the fork if desired :)

TheNitek commented:
Any other stability issues then the ones listed as issues on github? I'd be glad to help solve them, if that means I can get rid of my "SDFat+ID3Skip"-Fork ;-)

ladyada commented:
just the ones in github - basically something weird with ESP8266 or ESP32 - we aren't 100% sure what it is, we keep redoing our buffer code and it *seems* to get better!

TheNitek commented:
Any chance to get this into the master anytime soon? I'd be glad to help in any way possible.

ladyada commented:
sorry its been so long that its not longer synced to the current fork - do you want to try to make your own PR with this code based on the current master branch?

leftCoast commented:
OMG! I was thinking about this just this AM! I can grab the new code and
redo it. (I did the original) if you like.

-jim lee

On Sun, December 30, 2018 10:43 am, ladyada wrote:
> sorry its been so long that its not longer synced to the current fork -
> do you want to try to make your own PR with this code based on the
> current master branch?
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/adafruit/Adafruit_VS1053_Library/pull/23#issuecomment-4
> 50579197

ladyada commented:
@leftCoast heya wanna try the new PR?

leftCoast commented:
Sure, give me a day or so to swap over some hardware.

-jim lee

On Sun, December 30, 2018 2:35 pm, ladyada wrote:
> @leftCoast heya wanna try the new PR?
>
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/adafruit/Adafruit_VS1053_Library/pull/23#issuecomment-4
> 50591411

leftCoast commented:
Starting to bring my MP3 player up to date. (With the original library.) My graphics stuff has been all torn apart and rewritten so its going to take a bit. Didn't want you-all to think I'd forgotten. Soon as I get it up and running I'll swap in the new library and see how it all works.

leftCoast commented:
Ok, that was, as always, rougher that I expected. mp3 player is up and running on my original mp3 code & new interface code. Issue I worry about is this is running on the teensy 3.2 and they like using their "special" version and I currently have no Idea what makes it special. As I type i'm listening to the player run my original "non teensy" version seems fine. Off to try out TheNitek's branch.

P.S. TheNitek, how much did you change from my original? Or was it just a cut & past drop in thing? (I ask with hope in my voice.)

-jim lee

leftCoast commented:
Ok, as far as I can tell? It works wonderful. Also, TheNitek's one line version of "Is it an .mp3 file?" is way better than mine. Very clean and to the point. 
It compiles on my system (For teensy) with no warnings or errors. Runs exactly like the others as far as I can tell. Do you need this tested on the shield as well? I may be able to do a command line version to test that. (My GUI library no longer fits on a standard Arduino.)

There is another problem all the versions have though.. It has to do with volume setting. If you are watching a pot to set your volume during loop(), afterwards your processor runs VERY slow. Redraws are like molasses. The less often you call setVolume(), say every 100 ms, the less chance you have of seeing this happen. Or maybe you all already knew this?

-jim lee

TheNitek commented:
It was copy an paste + minimizing isMP3() and making it static, so not to many changes.

Maybe it makes sense to offload the volume problem into another issue. Do you call setVolume() during every loop()? Maybe that's causing to much SPI-Communication.

leftCoast commented:
I agree volume does NOT belong here. I just wanted to note it because it came up and eyes were on the issue. And, when I discover its all my fault because its something I did in my code, I'll be really embarrassed.

I ended up putting the volume call on a 100 ms timer. It helped somewhat. Still happens more than not though.

Another note : I do NOT use the interrupt version. As I notice TheNitek does, I just keep it fed through my loop() function. I almost never use delay, so its really not a problem. And, its so much simpler and easier to understand.

leftCoast commented:
OK, tried some more stuff. I updated all my Adafruit files to tonight's fresh copies. I had to tweak Adafruit_VS1053.h & SPIFTF.h

Basically 

```
#ifndef RwReg								// Yes this is my hack. - jim lee
 typedef volatile uint8_t RwReg;
#endif										// Other end of it.
```

In both files so they wouldn't conflict with one another.

I also added about 100 or so  .mp3 files to my player fresh from iTunes. So I have a mix of old and new.

The "new" files had no noticeable delay before starting. I though "Wow! All fixed we wasted our time." BUT the "old" files still had a delay. I thought they used to have about 20 second delays before. Now they are around 13 seconds. So the jump tags are still a good idea.

Also..

ALL my slow processor issues are gone! I can hammer the volume and after playback stops, everything still goes speedy quick! Very happy!

So, I think Apple is doing a better job of not leaving stuff in the files. And, these new libraries seem to work faster (for me) withe less issues. Except for the compiling error. But Im sure someone can address that.

I think that about wraps it up? You need anything else?

-jim lee

ladyada commented:
just lemme know when y'all agree this is ready :)

TheNitek commented:
@leftCoast if I understand correctly the RwReg fix is not related to this pull request, right? In other words you tested the rebased ID3 jumper implementation and it works fine, correct?

leftCoast commented:
@TheNitek Correct. The RwReg issue is independant from the pull request.  After I got the libraries to compile together, I rolled the jump tags into their current lib version and everything works fine. 

@ladyada Far as I can tell, works fine. The RwReg thing is a different issue.

TheNitek commented:
Sounds like this #49 can be merged and this one closed :-)

ladyada commented:
OK closing this one - we still need someone to test the other PR on another chip other than ESP8266 - say a samd21 or atmega328!
